src/registry/artifactory.rs

```diff
@@ -26,191 +27,166 @@ pub struct Artifactory {
 }
 
 impl Artifactory {
-    pub fn new(registry: RegistryUri, credentials: &Credentials) -> eyre::Result<Self> {
+    /// Creates a new instance of an Artifactory registry client
+    pub fn new(registry: RegistryUri, credentials: &Credentials) -> miette::Result<Self> {
         Ok(Self {
             registry: registry.clone(),
             token: credentials.registry_tokens.get(&registry).cloned(),
             client: reqwest::Client::builder()
                 .redirect(reqwest::redirect::Policy::none())
-                .build()?,
+                .build()
+                .into_diagnostic()?,
         })
     }
 
+    fn new_request(&self, method: Method, url: Url) -> RequestBuilder {
+        let mut request_builder = RequestBuilder::new(self.client.clone(), method, url);
+
+        if let Some(token) = &self.token {
+            request_builder = request_builder.auth(token.clone());
+        }
+
+        request_builder
+    }
+
     /// Pings artifactory to ensure registry access is working
-    pub async fn ping(&self) -> eyre::Result<()> {
+    pub async fn ping(&self) -> miette::Result<()> {
         let repositories_url: Url = {
             let mut uri = self.registry.to_owned();
             let path = &format!("{}/api/repositories", uri.path());
             uri.set_path(path);
             uri.into()
         };
 
-        let mut request = self.client.get(repositories_url);
-
-        if let Some(token) = &self.token {
-            request = request.bearer_auth(token);
-        }
-
-        let response = request.send().await?;
-
-        let status = response.status();
+        self.new_request(Method::GET, repositories_url)
+            .send()
+            .await
+            .map(|_| ())
+            .map_err(miette::Report::from)
+    }
 
-        if !status.is_success() {
-            tracing::info!("Artifactory response header: {:?}", response);
-        }
+    /// Downloads a package from artifactory
+    pub async fn download(&self, dependency: Dependency) -> miette::Result<Package> {
+        let artifact_url = resolve_url(dependency.clone())?;
 
-        ensure!(status.is_success(), "Failed to ping artifactory");
+        tracing::debug!("Hitting download URL: {artifact_url}");
 
-        tracing::debug!("Pinging artifactory succeeded");
+        let response = self.new_request(Method::GET, artifact_url).send().await?;
 
-        Ok(())
+        convert_to_package(response)
```

Comment:
Hm, i have to say i dislike this, also the logic isnt reused right? Cant we just inline the logic for creating a pkg from the request?

Reply:
Two reasons for this refactor:
1. Splits out a self-contains unit of behaviour, making the download implementation much easier to understand.
2. Makes it less likely people will end up re-implementing the conversion from response to package and use a non-validated response.
Happy to revert to inline, but this was inspired by your suggestion to prevent people from using raw bytes to create Packages.